### ISSUE TITLE
roachtest: clearrange/checks=false/rangeTs=false failed waiting for merges to complete

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#106934]
roachtest.clearrange/checks=false/rangeTs=false [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10797165?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10797165?buildTab=artifacts#/clearrange/checks=false/rangeTs=false) on master @ [dbe8511fae8fca21562fdde5c240b1f7d06ef582](https://github.com/cockroachdb/cockroach/commits/dbe8511fae8fca21562fdde5c240b1f7d06ef582):

```
(test_runner.go:1075).runTest: test timed out (6h30m0s)
(monitor.go:137).Wait: monitor failure: monitor command failure: unexpected node event: 1: dead (exit status 137)
test artifacts and logs in: /artifacts/clearrange/checks=false/rangeTs=false/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=true</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*clearrange/checks=false/rangeTs=false.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-29499

This test timed out waiting for range merges to complete after dropping a table:
```

11:21:04 clearrange.go:168: test worker status: dropping table
11:21:04 clearrange.go:177: test worker status: computing number of ranges
11:21:04 clearrange.go:180: test worker status: dropping bank table
11:21:05 clearrange.go:192: test worker status: waiting for ~1907 merges to complete (and for at least 45m0s to pass)
11:21:05 clearrange.go:209: test worker status: waiting for ~1907 merges to complete (and for at least 44m59.744433686s to pass)
11:26:05 clearrange.go:209: test worker status: waiting for ~1869 merges to complete (and for at least 39m59.721759174s to pass)
11:31:05 clearrange.go:209: test worker status: waiting for ~1789 merges to complete (and for at least 34m59.72293405s to pass)
11:36:05 clearrange.go:209: test worker status: waiting for ~1781 merges to complete (and for at least 29m59.721224578s to pass)
11:41:05 clearrange.go:209: test worker status: waiting for ~1781 merges to complete (and for at least 24m59.706561511s to pass)
11:46:05 clearrange.go:209: test worker status: waiting for ~113 merges to complete (and for at least 19m59.763920766s to pass)
11:51:05 clearrange.go:209: test worker status: waiting for ~37 merges to complete (and for at least 14m59.770363455s to pass)
11:56:05 clearrange.go:209: test worker status: waiting for ~14 merges to complete (and for at least 9m59.769903524s to pass)
12:01:05 clearrange.go:209: test worker status: waiting for ~5 merges to complete (and for at least 4m59.771085603s to pass)
12:06:05 clearrange.go:209: test worker status: waiting for ~4 merges to complete (and for at least -225.414291ms to pass)
12:11:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -5m0.227645442s to pass)
12:16:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -10m0.234026284s to pass)
12:21:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -15m0.293525415s to pass)
12:26:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -20m0.222004048s to pass)
12:31:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -25m0.224342888s to pass)
12:36:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -30m0.22448796s to pass)
12:41:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -35m0.218643705s to pass)
12:46:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -40m0.229399418s to pass)
12:51:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -45m0.22595478s to pass)
12:56:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -50m0.222386644s to pass)
13:01:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -55m0.2223217s to pass)
13:06:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h0m0.223257238s to pass)
13:11:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h5m0.227596885s to pass)
13:16:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h10m0.222535247s to pass)
13:21:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h15m0.226086517s to pass)
13:26:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h20m0.222587784s to pass)
13:31:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h25m0.228270073s to pass)
13:36:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h30m0.226161704s to pass)
13:41:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h35m0.228131134s to pass)
13:46:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h40m0.227593852s to pass)
13:51:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h45m0.233837153s to pass)
13:56:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h50m0.233314185s to pass)
14:01:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -1h55m0.236271427s to pass)
14:06:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h0m0.232145131s to pass)
14:11:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h5m0.233131284s to pass)
14:16:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h10m0.2349026s to pass)
14:21:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h15m0.236264932s to pass)
14:26:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h20m0.23287477s to pass)
14:31:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h25m0.234926837s to pass)
14:36:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h30m0.237805905s to pass)
14:41:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h35m0.237914384s to pass)
14:46:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h40m0.237737948s to pass)
14:51:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h45m0.239181577s to pass)
14:56:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h50m0.242637729s to pass)
15:01:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -2h55m0.237795081s to pass)
15:06:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h0m0.238128929s to pass)
15:11:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h5m0.240077996s to pass)
15:16:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h10m0.243662783s to pass)
15:21:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h15m0.236465336s to pass)
15:26:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h20m0.238152823s to pass)
15:31:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h25m0.238193296s to pass)
15:36:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h30m0.241649507s to pass)
15:41:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h35m0.237804157s to pass)
15:46:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h40m0.240259524s to pass)
15:51:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h45m0.243679255s to pass)
15:56:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h50m0.244683655s to pass)
16:01:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -3h55m0.246102155s to pass)
16:06:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h0m0.250570133s to pass)
16:11:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h5m0.2489413s to pass)
16:16:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h10m0.248003214s to pass)
16:21:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h15m0.248887762s to pass)
16:26:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h20m0.249109028s to pass)
16:31:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h25m0.246024419s to pass)
16:36:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h30m0.248277441s to pass)
16:41:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h35m0.245814925s to pass)
16:46:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h40m0.250521297s to pass)
16:51:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h45m0.248721388s to pass)
16:56:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h50m0.252732454s to pass)
17:01:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -4h55m0.254247578s to pass)
17:06:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -5h0m0.254574392s to pass)
17:11:05 clearrange.go:209: test worker status: waiting for ~2 merges to complete (and for at least -5h5m0.25641485s to pass)
17:12:21 test_impl.go:349: test failure #1: full stack retained in failure_1.log: (test_runner.go:1075).runTest: test timed out (6h30m0s)
```

**Comment by RaduBerinde** [comment]
Observations by @andrewbaptist:

Looking at this more - it is stuck with this:
```
 E230706 12:09:34.182913 20241086 kv/kvserver/queue.go:1146 ⋮ [T1,n7,merge,s7,r63/11:‹/Table/{62-106/1/578…}›] 1764  removing learners from r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):       5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536]: change replicas of r1869 failed: descriptor changed: [expected] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5LEARNER, (n1,s1):3, (n6,s6):   7, next=8, gen=536] != [actual] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5VOTER_DEMOTING_LEARNER, (n1,s1):3, (n6,s6):7VOTER_INCOMING, next=8, gen=535]
```
I don't understand why the descriptor on n7 is stuck at gen 535 "forever" when the expected descriptor is 536. 

Here are all the logs related to it:
```
I230706 11:11:38.402154 293775 13@kv/kvserver/replica_raftstorage.go:503 ⋮ [T1,n7,s7,r1869/4:{-}] 1553  applying INITIAL snapshot 8dd5a538 from (n1,s1):3 at applied index 13
 I230706 11:11:38.555247 425 kv/kvserver/store_remove_replica.go:157 ⋮ [T1,n7,s7,r1869/4:‹/Table/106/1/5{78385…-85900…}›,raft] 1613  removing replica r1869/4
 E230706 12:06:24.347239 19037178 kv/kvserver/queue.go:1146 ⋮ [T1,n7,merge,s7,r1376/4:‹/Table/106/1/{168649…-578385…}›] 1720  removing learners from r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,  s2):5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536]: change replicas of r1869 failed: descriptor changed: [expected] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5LEARNER, (n1,s1):3, (n6,   s6):7, next=8, gen=536] != [actual] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5VOTER_DEMOTING_LEARNER, (n1,s1):3, (n6,s6):7VOTER_INCOMING, next=8, gen=535]
```
The confusing part is that n7 is not part of this range at all - I think it is sending this because it holds the "LHS" ranges. Here is a message from store 8 also not part of the range.
```
 E230706 12:00:19.021719 13737824 kv/kvserver/queue.go:1146 ⋮ [T1,n8,merge,s8,r1400/8:‹/Table/106/1/{430713…-578385…}›] 1630  removing learners from r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,  s2):5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536]: change replicas of r1869 failed: descriptor changed: [expected] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5LEARNER, (n1,s1):3, (n6,   s6):7, next=8, gen=536] != [actual] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5VOTER_DEMOTING_LEARNER, (n1,s1):3, (n6,s6):7VOTER_INCOMING, next=8, gen=535]
```

This is from n3:
```
 E230706 12:06:20.690704 16383067 kv/kvserver/queue.go:1146 ⋮ [T1,n3,merge,s3,r2136/1:‹/Table/106/1/{385236…-578385…}›] 1586  removing learners from r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,  s2):5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536]: change replicas of r1869 failed: descriptor changed: [expected] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5LEARNER, (n1,s1):3, (n6,   s6):7, next=8, gen=536] != [actual] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5VOTER_DEMOTING_LEARNER, (n1,s1):3, (n6,s6):7VOTER_INCOMING, next=8, gen=535]
```

Node 6:
```
 W230706 17:11:16.501417 25048461 kv/kvserver/replica_command.go:3361 ⋮ [T1,n6,s6,r1869/7:‹/Table/106/1/578{38509-90304}›,*kvpb.AdminRelocateRangeRequest] 5684  removing learners from r1869:‹/Table/106/1/ 578{38509-90304}› [(n3,s3):6, (n2,s2):5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536]: change replicas of r1869 failed: descriptor changed: [expected] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6,   (n2,s2):5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536] != [actual] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5VOTER_DEMOTING_LEARNER, (n1,s1):3, (n6,s6):7VOTER_INCOMING, next=8,         gen=535]
```
Node 2
```
 E230706 11:59:56.921871 16280700 kv/kvserver/queue.go:1146 ⋮ [T1,n2,merge,s2,r1729/1:‹/Table/106/1/5{07980…-78385…}›] 1705  removing learners from r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,   s2):5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536]: change replicas of r1869 failed: descriptor changed: [expected] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5LEARNER, (n1,s1):3, (n6,   s6):7, next=8, gen=536] != [actual] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5VOTER_DEMOTING_LEARNER, (n1,s1):3, (n6,s6):7VOTER_INCOMING, next=8, gen=535]
```
Node 1:
```
 E230706 11:48:17.898078 6153110 kv/kvserver/queue.go:1146 ⋮ [T1,n1,merge,s1,r1303/2:‹/Table/106/1/5{59546…-78385…}›] 1540  removing learners from r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,    s2):5LEARNER, (n1,s1):3, (n6,s6):7, next=8, gen=536]: change replicas of r1869 failed: descriptor changed: [expected] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5LEARNER, (n1,s1):3, (n6,   s6):7, next=8, gen=536] != [actual] r1869:‹/Table/106/1/578{38509-90304}› [(n3,s3):6, (n2,s2):5VOTER_DEMOTING_LEARNER, (n1,s1):3, (n6,s6):7VOTER_INCOMING, next=8, gen=535]
```
So apparently no-one has gen 536?

There is this one ominous message on n6 (the leader at this point):
```
 teamcity-10797165-1688622302-45-n10cpu16-0006> I230706 11:42:58.418838 9768571 13@kv/kvserver/replica_raft.go:402 ⋮ [T1,n6,s6,r1869/7:‹×›] 376 proposing LEAVE_JOINT: after=[(n3,s3):6 (n2,s2):5LEARNER     (n1,s1):3 (n6,s6):7] next=8
 teamcity-10797165-1688622302-45-n10cpu16-0006> E230706 11:42:58.432814 9768066 kv/kvserver/queue.go:1146 ⋮ [T1,n6,merge,s6,r1870/2:‹×›] 387 while carrying out changes [{ADD_VOTER n6,s6} {REMOVE_VOTER n2, s2}]: change replicas of r1869 failed: ‹×›:609: config change rejected by raft; please retry
```


This is likely caused by https://github.com/cockroachdb/cockroach/pull/106267

**Comment by andrewbaptist** [comment]
Marking this as closed since it has not re-occurred since #106267 merged.
